idseq_dag/steps/fetch_tax_info.py

```diff
@@ -0,0 +1,152 @@
+import json
+import threading
+import re
+import time
+import wikipedia
+from Bio import Entrez
+import idseq_dag.util.command as command
+import idseq_dag.util.log as log
+import idseq_dag.util.s3 as s3
+from idseq_dag.engine.pipeline_step import PipelineStep
+
+
+class PipelineStepFetchTaxInfo(PipelineStep):
+    '''
+        fetch tax info based on a list
+    '''
+    def run(self):
+        '''
+            1. fetch the taxid -> wikipedia link mappin
+            2. fetch wikipedia content
+            3. store everything
+        '''
+        taxid_list = self.input_files_local[0][0]
+        (taxid2wiki, taxid2desc) = self.output_files_local()
+
+        taxid2wikidict = {}
+        Entrez.email = self.additional_attributes.get("entrez_email", "yunfang@chanzuckerberg.com")
+
+        if s3.check_s3_presence(self.s3_path(taxid2wiki)):
+            # generated
+            taxid2wiki = s3.fetch_from_s3(self.s3_path(taxid2wiki), taxid2wiki)
+            with open(taxid2wiki, "r") as taf:
+                for line in taf:
+                    (key, val) = line.rstrip("\n").split("\t")
+                    taxid2wikidict[key] = val
+        else:
+            num_threads = self.additional_attributes.get("threads", 16)
+            batch_size = self.additional_attributes.get("batch_size", 100)
+            self.fetch_ncbi_wiki_map(num_threads, batch_size, taxid_list, taxid2wikidict)
+            # output the data
+            with open(taxid2wiki, 'w') as taxidoutf:
+                for taxid, wikiurl in taxid2wikidict.items():
+                    taxidoutf.write(f"{taxid}\t{wikiurl}\n")
+
+        # output dummay for actual wiki content for now
+        taxid2wikicontent = {}
+        self.fetching_wiki_content(taxid2wikidict, taxid2wikicontent)
+
+        with open(taxid2desc, 'w') as desc_outputf:
+            json.dump(taxid2wikicontent, desc_outputf)
+
+    @staticmethod
+    def fetching_wiki_content(taxid2wikidict, taxid2wikicontent):
+        threads = []
+        semaphore = threading.Semaphore(64) # 4x the threads
```

Comment:
maybe you could also use the additional_attributes["threads"] in this function?

Reply:
yep.